File: runtime/src/transaction/tags.rs
//! Transaction tags.
use crate::common::crypto::hash::Hash;

/// Tag is a key/value pair of arbitrary byte blobs with runtime-dependent
/// semantics which can be indexed to allow easier lookup of blocks and
/// transactions on runtime clients.
#[derive(Clone, Debug, Default)]
pub struct Tag {
    /// The tag key.
    pub key: Vec<u8>,
    /// The tag value.
    pub value: Vec<u8>,
    /// The hash of the transaction that emitted the tag.
    pub tx_hash: Hash,
}

/// A set of tags.
pub type Tags = Vec<Tag>;

impl Tag {
    /// Create a new tag.
    ///
    /// The transaction hash is not initialized.
    pub fn new(key: Vec<u8>, value: Vec<u8>) -> Self {
        Self {
            key,
            value,
            ..Default::default()
        }
    }
}

File: go/runtime/transaction/tags.go
package transaction

import "github.com/oasisprotocol/oasis-core/go/common/crypto/hash"

// TagBlockTxHash is the hash used for block emitted tags not tied to a specific
// transaction.
var TagBlockTxHash = hash.Hash([32]byte{})

// Tag is a key/value pair of arbitrary byte blobs with runtime-dependent
// semantics which can be indexed to allow easier lookup of transactions
// on runtime clients.
type Tag struct {
	// Key is the tag key.
	Key []byte
	// Value is the tag value.
	Value []byte
	// TxHash is the hash of the transaction that emitted the tag.
	TxHash hash.Hash
}

// Tags is a set of tags.
type Tags []Tag
